File: app/src/main/java/at/bitfire/icsdroid/LocalCalendar.java
package at.bitfire.icsdroid;

import android.accounts.Account;
import android.content.ContentProviderClient;
import android.content.ContentProviderOperation.Builder;
import android.content.ContentValues;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.provider.CalendarContract;
import android.util.Log;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import at.bitfire.ical4android.AndroidCalendar;
import at.bitfire.ical4android.AndroidCalendarFactory;
import at.bitfire.ical4android.AndroidEvent;
import at.bitfire.ical4android.AndroidEventFactory;
import at.bitfire.ical4android.CalendarStorageException;

public class LocalCalendar extends AndroidCalendar {

    protected String url, eTag, lastModified;

    private LocalCalendar(Account account, ContentProviderClient providerClient, AndroidEventFactory eventFactory, long id) {
        super(account, providerClient, eventFactory, id);
    }

    public static LocalCalendar[] findAll(Account account, ContentProviderClient provider) throws CalendarStorageException {
        return (LocalCalendar[])AndroidCalendar.findAll(account, provider, Factory.FACTORY);
    }


    @Override
    protected void populate(ContentValues info) {
        super.populate(info);
        url = info.getAsString(CalendarContract.Calendars.NAME);
        eTag = info.getAsString(CalendarContract.Calendars.CAL_SYNC1);
        lastModified = info.getAsString(CalendarContract.Calendars.CAL_SYNC2);
    }

    public LocalEvent[] findByUID(String uid) throws CalendarStorageException {
        return (LocalEvent[])query(AndroidEvent.COLUMN_UID + "=?", new String[] { uid });
    }

    public int retainByUID(String[] uids) throws CalendarStorageException {
        String[] escapedUIDs = new String[uids.length];
        int idx = 0;
        for (String uid : uids)
            escapedUIDs[idx++] = DatabaseUtils.sqlEscapeString(uid);
        return delete(LocalEvent.COLUMN_UID + " NOT IN (" + StringUtils.join(escapedUIDs, ",") +")", null);
    }


    public static class Factory implements AndroidCalendarFactory {
        public static final Factory FACTORY = new Factory();

        @Override
        public LocalCalendar newInstance(Account account, ContentProviderClient provider, long id) {
            return new LocalCalendar(account, provider, LocalEvent.LocalEventFactory.FACTORY, id);
        }

        @Override
        public LocalCalendar[] newArray(int size) {
            return new LocalCalendar[size];
        }

    }

}

File: app/src/main/java/at/bitfire/icsdroid/db/sync/LocalCalendar.kt
package at.bitfire.icsdroid.db.sync

import android.accounts.Account
import android.content.ContentProviderClient
import android.content.ContentValues
import android.provider.CalendarContract.Calendars
import at.bitfire.ical4android.AndroidCalendar
import at.bitfire.ical4android.AndroidCalendarFactory

class LocalCalendar private constructor(
    account: Account,
    provider: ContentProviderClient,
    id: Long,
): AndroidCalendar<LocalEvent>(account, provider, LocalEvent.Factory(), id) {

    companion object {
        const val COLUMN_ETAG = Calendars.CAL_SYNC1
        const val COLUMN_LAST_MODIFIED = Calendars.CAL_SYNC4
        const val COLUMN_LAST_SYNC = Calendars.CAL_SYNC5
        const val COLUMN_ERROR_MESSAGE = Calendars.CAL_SYNC6

        /**
         * Stores if the calendar's embedded alerts should be ignored.
         * @since 20221202
         */
        const val COLUMN_IGNORE_EMBEDDED = Calendars.CAL_SYNC8

        /**
         * Stores the default alarm to set to all events in the given calendar.
         * @since 20221202
         */
        const val COLUMN_DEFAULT_ALARM = Calendars.CAL_SYNC7
    }

    /** URL of iCalendar file */
    var url: String? = null
    /** iCalendar ETag at last successful sync */
    var eTag: String? = null

    /** iCalendar Last-Modified at last successful sync (or 0 for none) */
    var lastModified = 0L
    /** time of last sync (0 if none) */
    var lastSync = 0L
    /** error message (HTTP status or exception name) of last sync (or null) */
    var errorMessage: String? = null

    /** Setting: whether to ignore alarms embedded in the Webcal */
    var ignoreEmbeddedAlerts: Boolean? = null
    /** Setting: Shall a default alarm be added to every event in the calendar? If yes, this
     *  field contains the minutes before the event. If no, it is *null*. */
    var defaultAlarmMinutes: Long? = null

    override fun populate(info: ContentValues) {
        super.populate(info)
        url = info.getAsString(Calendars.NAME)

        eTag = info.getAsString(COLUMN_ETAG)
        info.getAsLong(COLUMN_LAST_MODIFIED)?.let { lastModified = it }

        info.getAsLong(COLUMN_LAST_SYNC)?.let { lastSync = it }
        errorMessage = info.getAsString(COLUMN_ERROR_MESSAGE)

        info.getAsBoolean(COLUMN_IGNORE_EMBEDDED)?.let { ignoreEmbeddedAlerts = it }
        info.getAsLong(COLUMN_DEFAULT_ALARM)?.let { defaultAlarmMinutes = it }
    }

    class Factory: AndroidCalendarFactory<LocalCalendar> {
        override fun newInstance(account: Account, provider: ContentProviderClient, id: Long): LocalCalendar =
            LocalCalendar(account, provider, id)
    }

}
